core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -639,7 +638,9 @@ protected S getNext() {
          Publisher<S> innerPublisher = localPublisher(segmentsToFilter, ch,
                keysToExclude == null ? Collections.emptySet() :
                      (segmentsToFilter == null ? IntStream.range(0, ch.getNumSegments()) : segmentsToFilter.intStream())
-                           .mapToObj(i -> keysToExclude.apply(i).stream()).flatMap(Function.identity()).collect(Collectors.toSet()),
+                           .mapToObj(i -> keysToExclude.apply(i).stream())
+                           .flatMap(Function.identity())
+                           .collect(Collectors.toSet()),
```

Comment:
Unfortunately, this isn't possible as we have to send the excluded keys to remote node as well. So it is best to put it in the set. The goal is to complete as many segments as possible so that this collection isn't that big.

The only way to solve this is by doing iteration over the container/store 1 segment at a time. This requires all the segmented stuff to be in first though. This way we only have a small subset of segments with excluded keys.

Reply:
Ok, so we'll have to wait for the segmentation stuff to be finished :)

To go a bit off-topic, what's the benefit of keeping the keys in separate sets by segment if we're going to merge the sets to the remote node anyway? If all the segments primary-owned by one node are iterated in a single go, they will finish at approximately the same time, so the lifetime of the sets also seems very similar.

Reply:
In the happy path yes they should all be closed at the same time. However if a rehash occurs then only a subset of segments will be completed. In which case we can throw away the ones that completed and only need to exclude the keys from the ones that haven't completed yet.

Reply:
Thanks for the explanation! Imagining a serialized `StreamRequestCommand` holding the same keys in memory yet again makes me a bit squeamish, especially with a store enabled, so I'm looking forward to iterating one segment at a time.